File: Updog.Domain/Vote/VotableEntity.cs
using System;

namespace Updog.Domain {
    /// <summary>
    /// A resource that can be voted. on.
    /// </summary>
    public abstract class VotableEntity : IUserEntity {
        #region Properties
        /// <summary>
        /// The unique ID of the entity.
        /// </summary>
        public int Id { get; set; }

        /// <summary>
        /// The user it belongs to.
        /// </summary>
        public int UserId { get; set; }

        /// <summary>
        /// The vote resource type it is.
        /// </summary>
        public abstract VotableEntityType EntityType { get; }

        /// <summary>
        /// How many upvotes the resource has recieved.
        /// </summary>
        public int Upvotes { get; set; }

        /// <summary>
        /// How many downvotes the resource has recieved.
        /// </summary>
        public int Downvotes { get; set; }
        #endregion

        #region Publics
        /// <summary>
        /// Add a new vote to the counts.
        /// </summary>
        /// <param name="vote">The type of vote to add.</param>
        public void AddVote(VoteDirection vote) {
            switch (vote) {
                case VoteDirection.Up:
                    Upvotes++;
                    break;
                case VoteDirection.Down:
                    Downvotes++;
                    break;
            }
        }

        /// <summary>
        /// Remove a vote from the couns.
        /// </summary>
        /// <param name="vote">The vote to remove.</param>
        public void RemoveVote(VoteDirection vote) {
            switch (vote) {
                case VoteDirection.Up:
                    Upvotes--;
                    break;
                case VoteDirection.Down:
                    Downvotes--;
                    break;
            }
        }
        #endregion
    }
}

File: Updog.Client/src/vote/common/votable-entity.ts
import { UserEntity } from '@/user/common/user-entity';
import { VoteResourceType } from '../domain/vote-resource-type';
import { VoteDirection } from '../domain/vote-direction';
import { Vote } from '../domain/vote';

/**
 * Base class for a resource that can be voted on.
 */
export abstract class VotableEntity extends UserEntity {
    /**
     * The current vote state of the user viewing it.
     */
    public abstract vote: Vote | null;

    /**
     * How many upvotes the resource has recieved.
     */
    public abstract upvotes: number;

    /**
     * How many downvotes the resource has recieved.
     */
    public abstract downvotes: number;

    /**
     * The type of vote resource it is.
     */
    public abstract voteResourceType: VoteResourceType;

    /**
     * The karma points the post has recieved (upvotes - downvotes).
     */
    get karma() {
        return this.upvotes - this.downvotes;
    }

    get upvoteRatio() {
        return (this.upvotes / (this.upvotes + this.downvotes)) * 100;
    }

    /**
     * Apply a vote to the entity.
     * @param direction The direction of the vote to apply.
     */
    public applyVote(direction: VoteDirection) {
        this.undoVote();

        switch (direction) {
            case VoteDirection.Up:
                this.vote = new Vote(VoteResourceType.Post, this.id, VoteDirection.Up);
                this.upvotes++;
                break;
            case VoteDirection.Neutral:
                this.vote = new Vote(VoteResourceType.Post, this.id, VoteDirection.Neutral);
                break;
            case VoteDirection.Down:
                this.vote = new Vote(VoteResourceType.Post, this.id, VoteDirection.Down);
                this.downvotes--;
                break;
        }
    }

    /**
     * Helper to undo a vote on the post.
     */
    private undoVote() {
        if (this.vote != null) {
            switch (this.vote.direction) {
                case VoteDirection.Up:
                    this.upvotes--;
                    break;
                case VoteDirection.Down:
                    this.downvotes--;
                    break;
            }

            this.vote = null;
        }
    }
}
